### PR TITLE
space leak wack-a-mole

### PR DESCRIPTION
**Pull request by buckie**
There's still one left though, it lives in `Messaging.ZMQ.sendProcess` and try as I might it out foxed me. 
- moving `messageReceiver` to `IO` fixed the major one
- `ZMQ_CONFLATE` got dropped during a merge and reinstalling it fixed another
- `zmq4-haskell` seems to have a space leak for recursive socket calls but the benefit from this may have been noise
- switching to `RWS.Strict` also seems to have helped too but may be noise

### CODE REVIEW

**Comment by buckie** [comment]
Update:
- changed `NodeID` from `{host, port}` to `{host,port,fullAddr}` and changes `nodeId2Addr` to just use `fullAddr`. Apparently ghc wasn't noticing that the `addrs` were the same and storing each.

`sendProcess` still leaks a tiny bit though, this trick just cut the amount in half. 

Hypothesis:  `ZMQ_CONFLATE` and ghc aren't getting along -- conflate means drop the cached message when a new one is given to ZMQ to send but GHC isn't collecting it

File: src/Juno/Types/State.hs [hunk 1]
```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Juno.Types.State
```
Comment (sirlensalot) [inline src/Juno/Types/State.hs]: "State" is a bad name, call it "Monad" or "Spec".

File: src/Juno/Types/State.hs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import System.Random (Random)
+
+import Juno.Types.Base
+import Juno.Types.Config
+import Juno.Types.Message
+import Juno.Types.Command
+import Juno.Types.Metric
+import Juno.Types.Log
+
+data Event = ERPC RPC
```
Comment (sirlensalot) [inline src/Juno/Types/State.hs]: Why is this in here??

File: src/Juno/Types/Base.hs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+deriving instance Eq Signature
+deriving instance Ord Signature
+instance Serialize Signature where
+  put (Sig s) = S.put s
+  get = Sig <$> (S.get >>= S.getByteString)
+
+instance Eq PublicKey where
+  b == b' = exportPublic b == exportPublic b'
+instance Ord PublicKey where
+  b < b' = exportPublic b < exportPublic b'
```
Comment (sirlensalot) [inline src/Juno/Types/Base.hs]: Just implement required method

File: src/Juno/Types/Base.hs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  parseJSON _ = mzero
+instance ToJSON (Map NodeID PublicKey) where
+  toJSON = toJSON . Map.toList
+instance FromJSON (Map NodeID PublicKey) where
+  parseJSON = fmap Map.fromList . parseJSON
+
+instance Eq PrivateKey where
+  b == b' = exportPrivate b == exportPrivate b'
+instance Ord PrivateKey where
+  b < b' = exportPrivate b < exportPrivate b'
```
Comment (sirlensalot) [inline src/Juno/Types/Base.hs]: Just implement required methods

**Pull request merged**